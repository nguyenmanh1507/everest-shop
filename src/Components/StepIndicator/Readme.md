StepIndicator example

```js
const steps = ['Shopping Cart', 'Payment', 'Order Complete'];

<StepIndicator steps={steps} currentStep={1} />
```
