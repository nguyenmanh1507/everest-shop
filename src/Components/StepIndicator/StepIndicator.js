// @flow

import React, { PureComponent, Fragment } from 'react'
import cx from 'classnames'
import { uniqueId, padStart } from 'lodash'

type Props = {
  currentStep: number,
  steps: Array<string>
}

class StepIndicator extends PureComponent<Props> {
  render() {
    const { currentStep, steps } = this.props

    return (
      <div className="step">
        {steps.map((step, index) => (
          <Fragment key={uniqueId('step')}>
            <div
              className={cx('step__checkpoint', {
                done: index + 1 <= currentStep
              })}
              data-test="step-checkpoint"
            >
              <div className="step__item">
                {padStart((index + 1).toString(), 2, '0')}
              </div>
              <div className="step__text">{step}</div>
            </div>
            {!(steps.length === index + 1) && (
              <div className="step__line" data-test="step-line">
                <div className="step__line-icon">
                  <i className="fa fa-caret-right" />
                </div>
              </div>
            )}
          </Fragment>
        ))}
      </div>
    )
  }
}

export default StepIndicator
