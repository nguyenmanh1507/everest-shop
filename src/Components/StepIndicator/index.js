// @flow

export { default } from './StepIndicator'
