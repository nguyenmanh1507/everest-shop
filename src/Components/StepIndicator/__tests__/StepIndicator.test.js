// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import forEach from 'lodash/forEach'

import StepIndicator from '../StepIndicator'

configure({ adapter: new Adapter() })

test('StepIndicator render properly', () => {
  let component = renderer.create(renderStepIndicator())

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // Step 2
  component = renderer.create(renderStepIndicator(2))
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // Step 3
  component = renderer.create(renderStepIndicator(3))
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('<StepIndicator>', () => {
  test('First checkpoint should has done class if currentStep = 1', () => {
    const wrapper = shallow(renderStepIndicator())

    expect(
      wrapper
        .find('[data-test="step-checkpoint"]')
        .first()
        .hasClass('done')
    ).toBeTruthy()
  })

  test('All checkpoint should has done class if currentStep = total steps', () => {
    const wrapper = shallow(renderStepIndicator(3))

    const checkpoints = wrapper
      .find('[data-test="step-checkpoint"]')
      .forEach(node => {
        expect(node.hasClass('done')).toBeTruthy()
      })
  })

  test('Checkpoints that has step number <= current step should has done class and ver vice', () => {
    let i = 1
    const stepNumber = 2
    const wrapper = shallow(renderStepIndicator(stepNumber))
    wrapper.find('[data-test="step-checkpoint"]').forEach(node => {
      if (i <= stepNumber) {
        expect(node.hasClass('done')).toBeTruthy()
      } else {
        expect(node.hasClass('done')).toBeFalsy()
      }
      i++
    })
  })

  test('Total step line should equal total step checkpoint - 1', () => {
    const wrapper = shallow(renderStepIndicator())

    const totalStepLine = wrapper.find('[data-test="step-line"]').length
    const totalStepCheckpoint = wrapper.find('[data-test="step-checkpoint"]')
      .length

    expect(totalStepLine).toBe(totalStepCheckpoint - 1)
  })
})

function renderStepIndicator(currentStep = 1) {
  const steps = ['Shopping Cart', 'Payment', 'Order Complete']

  return <StepIndicator steps={steps} currentStep={currentStep} />
}
