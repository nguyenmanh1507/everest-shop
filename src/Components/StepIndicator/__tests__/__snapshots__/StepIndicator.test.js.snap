// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepIndicator render properly 1`] = `
<div
  className="step"
>
  <div
    className="step__checkpoint done"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      01
    </div>
    <div
      className="step__text"
    >
      Shopping Cart
    </div>
  </div>
  <div
    className="step__line"
    data-test="step-line"
  >
    <div
      className="step__line-icon"
    >
      <i
        className="fa fa-caret-right"
      />
    </div>
  </div>
  <div
    className="step__checkpoint"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      02
    </div>
    <div
      className="step__text"
    >
      Payment
    </div>
  </div>
  <div
    className="step__line"
    data-test="step-line"
  >
    <div
      className="step__line-icon"
    >
      <i
        className="fa fa-caret-right"
      />
    </div>
  </div>
  <div
    className="step__checkpoint"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      03
    </div>
    <div
      className="step__text"
    >
      Order Complete
    </div>
  </div>
</div>
`;

exports[`StepIndicator render properly 2`] = `
<div
  className="step"
>
  <div
    className="step__checkpoint done"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      01
    </div>
    <div
      className="step__text"
    >
      Shopping Cart
    </div>
  </div>
  <div
    className="step__line"
    data-test="step-line"
  >
    <div
      className="step__line-icon"
    >
      <i
        className="fa fa-caret-right"
      />
    </div>
  </div>
  <div
    className="step__checkpoint done"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      02
    </div>
    <div
      className="step__text"
    >
      Payment
    </div>
  </div>
  <div
    className="step__line"
    data-test="step-line"
  >
    <div
      className="step__line-icon"
    >
      <i
        className="fa fa-caret-right"
      />
    </div>
  </div>
  <div
    className="step__checkpoint"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      03
    </div>
    <div
      className="step__text"
    >
      Order Complete
    </div>
  </div>
</div>
`;

exports[`StepIndicator render properly 3`] = `
<div
  className="step"
>
  <div
    className="step__checkpoint done"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      01
    </div>
    <div
      className="step__text"
    >
      Shopping Cart
    </div>
  </div>
  <div
    className="step__line"
    data-test="step-line"
  >
    <div
      className="step__line-icon"
    >
      <i
        className="fa fa-caret-right"
      />
    </div>
  </div>
  <div
    className="step__checkpoint done"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      02
    </div>
    <div
      className="step__text"
    >
      Payment
    </div>
  </div>
  <div
    className="step__line"
    data-test="step-line"
  >
    <div
      className="step__line-icon"
    >
      <i
        className="fa fa-caret-right"
      />
    </div>
  </div>
  <div
    className="step__checkpoint done"
    data-test="step-checkpoint"
  >
    <div
      className="step__item"
    >
      03
    </div>
    <div
      className="step__text"
    >
      Order Complete
    </div>
  </div>
</div>
`;
